torch/csrc/jit/script/compiler.cpp

```diff
@@ -72,14 +176,114 @@ struct to_ir {
+    return save_env;
+  }
+
   void emitIf(const If& stmt) {
-    // TODO: add support for control flow ops
-    throw ErrorReport(stmt) << "Control flow is not supported yet.";
+    Value* cond_value = emitExpr(stmt.cond(), 1)[0];
+
+    Node* n = def.graph->insertNode(def.graph->create(Symbol("If"), 0));
+    n->addInput(cond_value);
+    auto* true_block = n->addBlock();
+    auto* false_block = n->addBlock();
+
+    // Emit both blocks once to get the union of all mutated values
+    std::unordered_set<std::string> mutated_parent_values;
+    std::shared_ptr<Environment> save_true, save_false;
+    save_true = emitSingleIfBranch(
+        true_block, stmt.trueBranch(), &mutated_parent_values);
+    save_false = emitSingleIfBranch(
+        false_block, stmt.falseBranch(), &mutated_parent_values);
+
+    std::vector<std::string> sorted_mutations(
+        mutated_parent_values.begin(), mutated_parent_values.end());
+    std::sort(sorted_mutations.begin(), sorted_mutations.end());
+
+    // Register outputs in each block
+    environment_stack = save_true;
+    for (const auto& x : sorted_mutations) {
+      true_block->registerOutput(environment_stack->getVar(x, stmt));
+    }
+    environment_stack = save_false;
+    for (const auto& x : sorted_mutations) {
+      false_block->registerOutput(environment_stack->getVar(x, stmt));
+    }
+    environment_stack = environment_stack->next;
+
+    // Add op outputs
+    for (const auto& x : sorted_mutations) {
+      environment_stack->setVar(x, n->addOutput());
+    }
   }
 
   void emitWhile(const While& stmt) {
-    // TODO: add support for control flow ops
-    throw ErrorReport(stmt) << "Control flow is not supported yet.";
+    // Emits a loop operators conforming to the semantics specified at
+    // https://github.com/onnx/onnx/blob/master/docs/Operators.md#experimental-loop 
+    // TODO: implement scan_outputs
+
+    Value* trip_count_dummy = emitConst(0, "i")[0];
+    Value* cond_value = emitExpr(stmt.cond(), 1)[0];
+
+    Node* n = def.graph->insertNode(def.graph->create(Symbol("Loop"), 0));
+    n->addInput(trip_count_dummy);
+    n->addInput(cond_value);
+    auto* body_block = n->addBlock();
+    // Trip count required by spec. Since this is a while loop, we do not
+    // provide access to this from user code
+    // TODO: it seems like we should implement a `for` loop as well, otherwise
+    // we'll probably have to pattern match iteration number machinery in user
+    // code to conform to the spec
+    body_block->addInput(); // Iteration num
+    body_block->addInput(); // Condition
+    size_t skip_inputs_num = 2;
+
+    {
+      environment_stack =
+          std::make_shared<Environment>(body_block, environment_stack);
+      WithInsertPoint guard(*def.graph, body_block);
+      emitStatements(stmt.body());
+
+      // Also emit the conditional
+      Value *body_cond_value = emitExpr(stmt.cond(), 1)[0];
+      body_block->registerOutput(body_cond_value);
```

Comment:
It would feel more natural to postpone the conditional to the end of the loop, but I guess that's ok

Reply:
This is emitting after 260, so it is at the end of the loop, unless I'm misunderstanding you

Reply:
Ah, right. Never mind